packages/amplication-server/src/core/workspace/workspace.service.ts

```diff
@@ -339,6 +342,77 @@ export class WorkspaceService {
     return updatedInvitation;
   }
 
+  async redeemCoupon(
+    currentUser: User,
+    args: RedeemCouponArgs
+  ): Promise<boolean> {
+    const { account } = currentUser;
+
+    const coupons = await this.prisma.coupon.findMany({
+      where: {
+        code: args.data.code,
+        newUser: null,
+        redemptionAt: null,
+      },
+    });
+
+    if (!(coupons && coupons.length === 1)) {
+      throw new ConflictException(`coupon cannot be found or it has expired`);
+    }
+
+    const [coupon] = coupons;
+
+    if (coupon.expiration < new Date()) {
+      throw new ConflictException(`coupon is expired`);
+    }
+
+    const subscription = await this.subscriptionService.resolveSubscription(
+      currentUser.workspace.id
+    );
+
+    if (subscription.subscriptionPlan !== EnumSubscriptionPlan.Free) {
+      throw new ConflictException(
+        `The coupon cannot be applied on the current subscription`
+      );
+    }
+
+    //use stigg API to provision a new trial subscription
+    await this.billingService.provisionSubscription({
+      workspaceId: currentUser.workspace.id,
+      planId: "plan-amplication-pro-with-trial",
+      cancelUrl: "",
+      successUrl: "",
+      userId: currentUser.account.id,
+      billingPeriod: BillingPeriod.Monthly,
+      intentionType: "UPGRADE_PLAN",
+    });
+    await this.prisma.coupon.update({
+      where: {
+        id: coupon.id,
+      },
+      data: {
+        redemptionAt: new Date(),
+        newUser: {
+          connect: {
+            id: currentUser.id,
+          },
+        },
+      },
+    });
+
+    await this.analytics.track({
+      userId: account.id,
+      event: EnumEventType.RedeemCoupon,
+      properties: {
+        workspaceId: currentUser.workspace.id,
+        subscriptionPlan: coupon.subscriptionPlan,
+        durationMonths: coupon.durationMonths,
```

Comment:
this is currently not the true value (since the real value comes from stigg)

Reply:
true